src/parsers/io_stat.cpp

```diff
@@ -0,0 +1,70 @@
+/*
+ *  Copyright 2020-present Roman Karlstetter
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+
+#include "pfs/parsers.hpp"
+#include "pfs/utils.hpp"
+
+namespace pfs {
+namespace impl {
+namespace parsers {
+
+namespace {
+
+template <typename T>
+static void to_number(const std::string& value, T& out,
+                      utils::base base = utils::base::decimal)
+{
+    try
+    {
+        utils::stot(value, out, base);
+    }
+    catch (const std::invalid_argument& ex)
+    {
+        throw parser_error("Corrupted number - Invalid argument", value);
+    }
+    catch (const std::out_of_range& ex)
+    {
+        throw parser_error("Corrupted number - Out of range", value);
+    }
+}
+
+template <unsigned long io_stat::*member>
+static io_stat_parser::value_parser io_parser() {
```

Comment:
To prevent un-protecting `value_parser`, which violates the encapsulation IMHO, I suggest:
1. You define that as `static io_stat_parser::value_parser io_stat_parser::io_parser`, and then you don't need to un-protect the `value_parser` type, keeping it as a protected type. (This will require moving this to the header file)
2. Defining this method as a friend of the `io_stat_parser` class.

Reply:
My idea was to hide the implementation details in the cpp file and keep the header (and the io_stat_parser class definition) as clean as possible. To me, exposing the typedef feels better than polluting the interface with internals the user does not need to know. But I'll change it if you prefer improving encapsulation, of course.